users/users.go

```diff
@@ -0,0 +1,199 @@
+	FirstName      string         `json:"first_name,omitempty"`
+	LastName       string         `json:"last_name,omitempty"`
+	Tags           []string       `json:"tags,omitempty"`
+	Credentials    Credentials    `json:"credentials,omitempty"`
+	Metadata       Metadata       `json:"metadata,omitempty"`
+	CreatedAt      time.Time      `json:"created_at,omitempty"`
+	UpdatedAt      time.Time      `json:"updated_at,omitempty"`
+	UpdatedBy      string         `json:"updated_by,omitempty"`
+	Status         clients.Status `json:"status,omitempty"` // 1 for enabled, 0 for disabled
+	Role           clients.Role   `json:"role,omitempty"`   // 1 for admin, 0 for normal user
+	Permissions    []string       `json:"permissions,omitempty"`
+	ProfilePicture string         `json:"profile_picture,omitempty"`
+	DomainID       string         `json:"domain_id,omitempty"`
+}
+
+type Credentials struct {
+	Identity string `json:"identity,omitempty"` // username or generated login ID
+	Secret   string `json:"secret,omitempty"`   // password or token
+}
+
+type UsersPage struct {
+	clients.Page
+	Users []User
+}
+
+// Metadata represents arbitrary JSON.
+type Metadata map[string]interface{}
+
+// MembersPage contains page related metadata as well as list of members that
+// belong to this page.
+type MembersPage struct {
+	clients.Page
+	Members []User
+}
+
+// UserRepository struct implements the Repository interface.
+type UserRepository struct {
+	DB postgres.Database
+}
+
+//go:generate mockery --name Repository --output=./mocks --filename repository.go --quiet --note "Copyright (c) Abstract Machines"
+type Repository interface {
+	// RetrieveByID retrieves user by their unique ID.
+	RetrieveByID(ctx context.Context, id string) (User, error)
+
+	// RetrieveByIdentity retrieves user by its unique credentials.
+	RetrieveByIdentity(ctx context.Context, identity string) (User, error)
+
+	// RetrieveByUserName retrieves user by their user name
+	RetrieveByUserName(ctx context.Context, userName string) (User, error)
+
+	// RetrieveAll retrieves all users.
+	RetrieveAll(ctx context.Context, pm clients.Page) (UsersPage, error)
+
+	// Update updates the user name and metadata.
+	Update(ctx context.Context, user User) (User, error)
+
+	// UpdateTags updates the user tags.
+	UpdateTags(ctx context.Context, user User) (User, error)
+
+	// UpdateIdentity updates identity for user with given id.
+	UpdateIdentity(ctx context.Context, user User) (User, error)
+
+	// UpdateUserNames updates the User's names.
+	UpdateUserNames(ctx context.Context, user User) (User, error)
+
+	// UpdateProfilePicture updates profile picture for user with given ID.
+	UpdateProfilePicture(ctx context.Context, user User) (User, error)
+
+	// UpdateSecret updates secret for user with given identity.
+	UpdateSecret(ctx context.Context, user User) (User, error)
+
+	// UpdateRole updates role for user with given id.
+	UpdateRole(ctx context.Context, user User) (User, error)
```

Comment:
This looks like GraphQL, rather than REST. Can we have only one update and one for password reset? Eventually, we can have a separate method for profile picture update.